[WIP] Allow Wrapper Queries

alexanderkiel opened the pull request:
It's work in progress open for discussions. I don't like the `:wrapper-class` thing but don't know how to get the query from the factory. Also don't tested any remote stuff.

alexanderkiel committed:
Allow Wrapper Queries

Allow Wrapper Queries

anmonteiro commented on src/main/compassus/core.cljc:
```diff
@@ -17,7 +17,7 @@
 
 (defn- make-root-class
-  [{:keys [routes wrapper history]}]
+  [{:keys [routes wrapper wrapper-class history]}]
```
we can probably get at the query by invoking `om.next/react-type` on the wrapper thing and checking if it implements a query with `om.next/iquery?`

alexanderkiel replied:
wrapper is the component factory. as I understand it, a factory is really only a function. `(om/iquery? (om/react-type wrapper))` returns false.

anmonteiro commented:
I made a comment on the source.
I'd also like to see some tests for this, and possibly a devcards example

alexanderkiel commented:
Regarding tests and devcard example: yes for sure after the implementation makes sense.

alexanderkiel committed:
Only set query if one exists

alexanderkiel committed:
Queue wrapper data

anmonteiro commented:
let's do it like this:

```clojure
(defn ^boolean is-factory? [wrapper]
  (= (subs (pr-str wrapper) 0 20)
     "#object[om$next$self"))

wrapper-query (when (is-factory? wrapper)
                        (let [c (wrapper)
                              class (om/react-type c)]
                          (when (om/iquery? class)
                            (om/get-query class))))
```

alexanderkiel committed:
Calculate Query in Advance

alexanderkiel committed:
Specify Wrapper Query Directly

Previously the wrapper component specified the query like any other
component through om/IQuery. The problem with this is that all child
components of the wrapper don't have a valid om/full-query anymore
because there queries are not part of the wrapper query.

At the end not the wrapper has the query, the root component has the
query alongside the routing queries. We could rename the query in
something like static-query. This way we have the routing queries and
one static query and also static-props.

alexanderkiel commented:
What do you think about my commit d0824b8? I saw the problem with `om/full-query` while working with remotes. Is my idea to rename everything into static... a good one?

alexanderkiel committed:
Use Wrapper Query in Normalization

We need the wrapper query alongside the routing queries in
normalization. Otherwise part of the state which is only touched by the
wrapper query wouldn't be normalized.

anmonteiro commented:
this has been addressedn in b63876a806bb03b3e462754136bfcd24fdf2001b.
See #3 for an explanation.
